lib/stockman/publisher.rb

```diff
@@ -10,38 +10,55 @@ class << self
       #
       # @example Publish message
       #   Stockman::Publisher.publish 'set_sellable_product', "{\"sellable_id\":1,\"sellable_type\":\"variant\",\"product_id\":99,\"meta\":{\"foo\":\"bar\"}}"
-      def publish(routing_key, message)
-        connect! unless connected?
+      delegate :publish, :reset!, to: :publisher
 
-        exchange.publish message, routing_key: routing_key, persistent: true
+      def publisher
+        @publisher ||= new
       end
+    end
 
-      # Closes connection to RabbitMQ
-      def reset!
-        return unless connected?
+    def initialize
+      @mutex = Mutex.new
+    end
 
-        @bunny.close
-        @bunny = nil
-      end
+    def publish(routing_key, message)
+      ensure_connection!
 
-      private
+      exchange.publish message, routing_key: routing_key, persistent: true
+    end
+
+    # Closes connection to RabbitMQ
+    def reset!
+      return unless connected?
 
-      attr_reader :exchange
+      @bunny.close
+      @bunny = nil
+    end
 
-      def connect!
-        @bunny = Bunny.new(Stockman.amqp)
-        @bunny.start
+    private
 
-        if Stockman.exchange
-          @exchange = @bunny.create_channel.exchange(Stockman.exchange, type: :topic, durable: true)
-        else
-          @exchange = @bunny.create_channel.default_exchange
-        end
+    attr_reader :exchange
+
+    def ensure_connection!
+      @mutex.synchronize do
```

Comment:
```suggestion
      return if connected?
      @mutex.synchronize do
```

not to suppress thread benefits